### PR TITLE
Webhook implementation

### PR DESCRIPTION
**Pull request by TonyJabbour**

### CODE REVIEW

File: var/www/modules/hunter/templates/showTracker.html [hunk 1]
```diff
@@ -1,359 +1,400 @@
...
+                                            <td>
+                                                {% if tracker_metadata['last_seen'] %}
+                                                    {{ tracker_metadata['last_seen'][0:4] }}/
+                                                    {{ tracker_metadata['last_seen'][4:6] }}/
+                                                    {{ tracker_metadata['last_seen'][6:8] }}
+                                                {% endif %}
+                                            </td>
+                                            {% if tracker_metadata['webhook'] %}
+                                                <td>
+                                                    Turned ON
```
Comment (Terrtia) [inline var/www/modules/hunter/templates/showTracker.html:120]: I think we should show the full list (like mails)

File: var/www/modules/hunter/templates/edit_tracker.html [hunk 2]
```diff
@@ -43,7 +43,7 @@ <h5 class="card-title">Edit a Tracker</h5>
 									<div class="row">
 										<div class="col-12 col-xl-9">
 										  <div class="input-group mb-2 mr-sm-2">
-										    <div class="input-group-prepend">
+										    <div class="input-group-prepend"-->
```
Comment (Terrtia) [inline var/www/modules/hunter/templates/edit_tracker.html:46]: Small typo: ``--``

File: bin/lib/Tracker.py [hunk 3]
```diff
@@ -464,7 +474,7 @@ def create_tracker(tracker, tracker_type, user_id, level, tags, mails, descripti
     for source in sources:
         # escape source ?
         r_serv_tracker.sadd(f'tracker:sources:{tracker_uuid}', escape(source))
-
+    r_serv_tracker.sadd(f'tracker:webhook:{tracker_uuid}', webhook)
```
Comment (Terrtia) [inline bin/lib/Tracker.py:477]: How many web-hooks do you want to store ? one or a list ?

File: bin/trackers/Tracker_Regex.py [hunk 4]
```diff
@@ -93,7 +96,24 @@ def new_tracker_found(self, tracker, tracker_type, item):
...
+                                "itemId": item_id,
+                                "itemURL": self.full_item_url + item_id,
+                                "tracker": tracker,
+                                "itemSource": item_source,
+                                "itemDate": item_date,
+                                "tags": tags_to_add,
+                                "emailNotification": f'{mail_to_notify}',
+                                "trackerType": tracker_type
+                                }
+                response = requests.post(webhook_to_post, json=json_request)
```
Comment (Terrtia) [inline bin/trackers/Tracker_Regex.py:112]: Can you add a try/catch to handle exceptions raised by the requests library ?

File: bin/trackers/Tracker_Term.py [hunk 5]
```diff
@@ -135,8 +139,24 @@ def new_term_found(self, term, term_type, item_id, item_source):
...
+                                    "itemId": item_id,
+                                    "itemURL": self.full_item_url + item_id,
+                                    "term": term,
+                                    "itemSource": item_source,
+                                    "itemDate": item_date,
+                                    "tags": tags_to_add,
+                                    "emailNotification": f'{mail_to_notify}',
+                                    "trackerType": term_type
+                                    }
+                    response = requests.post(webhook_to_post, json=json_request)
```
Comment (Terrtia) [inline bin/trackers/Tracker_Term.py:155]: Can you add a try/catch to handle exceptions raised by the requests library ?

File: bin/trackers/Tracker_Yara.py [hunk 6]
```diff
@@ -96,10 +96,26 @@ def yara_rules_match(self, data):
...
+                            "itemId": item_id,
+                            "itemURL": self.full_item_url + item_id,
+                            "dataRule": data["rule"],
+                            "itemSource": item_source,
+                            "itemDate": item_date,
+                            "tags": tags_to_add,
+                            "emailNotification": f'{mail_to_notify}',
+                            "trackerType": "yara"
+                            }
+            response = requests.post(webhook_to_post, json=json_request)
```
Comment (Terrtia) [inline bin/trackers/Tracker_Yara.py:112]: Can you add a try/catch to handle exceptions raised by the requests library ?

File: bin/trackers/Tracker_Regex.py [hunk 7]
```diff
@@ -93,7 +96,24 @@ def new_tracker_found(self, tracker, tracker_type, item):
...
+                if response.status_code >= 400:
+                    raise Exception(f"Webhook request failed for {webhook_to_post}\nReason: {response.reason}")
```
Comment (Terrtia) [inline bin/trackers/Tracker_Regex.py:114]: I don't think we should stop the module if we can't reach the webhook.
Can you replace this exception by an error message ?

You can use this function:
```python
self.redis_logger.error(error_message)
```

File: bin/trackers/Tracker_Term.py [hunk 8]
```diff
@@ -135,8 +139,24 @@ def new_term_found(self, term, term_type, item_id, item_source):
...
+                    if response.status_code >= 400:
+                        raise Exception(f"Webhook request failed for {webhook_to_post}\nReason: {response.reason}")
```
Comment (Terrtia) [inline bin/trackers/Tracker_Term.py:157]: I don't think we should stop the module if we can't reach the webhook.
Can you replace this exception by an error message ?

You can use this function:
```python
self.redis_logger.error(error_message)
```

File: bin/trackers/Tracker_Yara.py [hunk 9]
```diff
@@ -96,10 +96,26 @@ def yara_rules_match(self, data):
...
+            if response.status_code >= 400:
+                raise Exception(f"Webhook request failed for {webhook_to_post}\nReason: {response.reason}")
```
Comment (Terrtia) [inline bin/trackers/Tracker_Yara.py:114]: I don't think we should stop the module if we can't reach the webhook.
Can you replace this exception by an error message ?

You can use this function:
```python
self.redis_logger.error(error_message)
```

**Review by Terrtia — CHANGES_REQUESTED** [review]
Hey @TonyJabbour !

Thank you very much! That’s great.

How many webhooks do you want to add ?

File: bin/lib/Tracker.py (hunk 3 above)
Comment (TonyJabbour) [reply bin/lib/Tracker.py:477]: my implementation 1 webgook / Tracker

File: var/www/modules/hunter/templates/showTracker.html (hunk 1 above)
Comment (TonyJabbour) [reply var/www/modules/hunter/templates/showTracker.html:120]: I did this at first, but the URL is soo long, made the table UI broken, so the only useful way was saying if ON or not showing it at all

**Review by Terrtia — APPROVED** [review]

**Comment by Terrtia** [comment]
Thanks a lot !

**Pull request merged**